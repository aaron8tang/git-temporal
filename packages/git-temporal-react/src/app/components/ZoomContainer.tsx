import React from 'react';
import { style } from 'app/styles';

export interface ZoomContainerProps {
  // The children are the menu content
  children: string | JSX.Element | JSX.Element[];
  onZoom?: (value: number) => void;
  style?: string | object;
}
const initialState = {
  zoom: 100,
};
type ZoomContainerState = Readonly<typeof initialState>;

const outerStyle = {
  _extends: 'fill',
  position: 'relative',
};

const containerStyle = {
  overflowX: 'scroll',
  height: '100%',
  width: '100%',
  position: 'relative',
};

const scrollContainerStyle = {
  height: '100%',
  // width: is computed based on state.zoom
};

const zoomSelectorStyle = {
  position: 'absolute',
  right: 18,
  bottom: -19,
  backgroundColor: style('backgroundColor')['backgroundColor'],
};

const availableZooms = [100, 200, 300, 500, 800, 1300];

export class ZoomContainer extends React.Component<
  ZoomContainerProps,
  ZoomContainerState
> {
  readonly state: ZoomContainerState = initialState;

  render() {
    const zoomStyle = {
      width: `${this.state.zoom || 100}%`,
    };
    return (
      <div style={style(outerStyle)}>
        <div style={style(this.props.style, containerStyle)}>
          <div style={style(scrollContainerStyle, zoomStyle)}>
            {this.props.children}
          </div>
        </div>
        <div style={style(zoomSelectorStyle)}>
          <label style={style('normalText')}>
            <select onChange={this.onZoomChange}>
              {this.renderZoomOptions()}
            </select>
          </label>
        </div>
      </div>
    );
  }

  private renderZoomOptions(): any {
    return availableZooms.map((zoom, index) => {
      const selected = zoom === this.state.zoom;
      return (
        <option key={index} value={zoom} selected={selected}>
          {zoom}% zoom
        </option>
      );
    });
  }

  private onZoomChange = evt => {
    const value = parseInt(evt.target.value, 10);
    this.setState({ zoom: value }, () => {
      this.props.onZoom(value);
    });
  };
}
