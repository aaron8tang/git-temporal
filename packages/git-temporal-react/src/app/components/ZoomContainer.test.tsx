import React from 'react';
import { shallow } from 'enzyme';

import { ZoomContainer } from './ZoomContainer';

describe('components/ZoomContainer', () => {
  describe('when rendered', () => {
    let wrapper;
    const onZoomMock = jest.fn();
    beforeAll(() => {
      wrapper = shallow(
        <ZoomContainer onZoom={onZoomMock}>
          <div style={{ width: '100%' }}>
            This is something contained in a ZoomContainer
          </div>
        </ZoomContainer>
      );
    });

    test('it should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('and then changing zoom prop', () => {
      wrapper.instance().onZoomChange({ target: { value: '400' } });
      expect(onZoomMock).toHaveBeenCalledTimes(1);
    });
  });
});
