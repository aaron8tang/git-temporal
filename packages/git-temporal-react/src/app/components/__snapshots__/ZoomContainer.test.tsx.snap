// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ZoomContainer when rendered it should match snapshot 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <HorizontalScroller>
    <div
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "width": "100%",
          }
        }
      >
        This is something contained in a ZoomContainer
      </div>
    </div>
  </HorizontalScroller>
  <div
    style={
      Object {
        "background": "transparent",
        "position": "absolute",
        "right": 18,
      }
    }
  >
    <label
      style={
        Object {
          "color": "#333333",
          "fontSize": 12,
          "fontWeight": "normal",
          "lineHeight": 1.35,
        }
      }
    >
      <select
        onChange={[Function]}
      >
        <option
          key="0"
          selected={true}
          value={100}
        >
          100% zoom
        </option>
        <option
          key="1"
          selected={false}
          value={200}
        >
          200% zoom
        </option>
        <option
          key="2"
          selected={false}
          value={300}
        >
          300% zoom
        </option>
        <option
          key="3"
          selected={false}
          value={500}
        >
          500% zoom
        </option>
        <option
          key="4"
          selected={false}
          value={800}
        >
          800% zoom
        </option>
        <option
          key="5"
          selected={false}
          value={1300}
        >
          1300% zoom
        </option>
      </select>
    </label>
  </div>
</div>
`;
